compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -136,6 +144,242 @@ emitLinkedTuningSpec(ModuleOp module, ArrayRef<NamedSequenceOp> specsToLink) {
+            (matcherName == specName)
+                ? SymbolRefAttr::get(builder.getContext(), newSpecName)
+                : matcherAttr);
+      }
+
+      for (auto actionAttr : foreachMatchOp.getActions()) {
+        StringRef actionName =
+            cast<SymbolRefAttr>(actionAttr).getRootReference();
+        updatedActions.push_back(
+            (actionName == specName)
+                ? SymbolRefAttr::get(builder.getContext(), newSpecName)
+                : actionAttr);
+      }
+
+      // Apply the updated matchers and actions.
+      foreachMatchOp.setMatchersAttr(builder.getArrayAttr(updatedMatchers));
+      foreachMatchOp.setActionsAttr(builder.getArrayAttr(updatedActions));
+    }
+    op.getOperation()->moveBefore(module.getBody(), module.getBody()->end());
+  }
+
+  // Step 3-b: Create a new NamedSequenceOp `__kernel_config` in the top-level
+  // module.
+  builder.setInsertionPointToEnd(module.getBody());
+  Location loc = module.getLoc();
+  Type anyOpType = builder.getType<transform::AnyOpType>();
+  FunctionType seqType =
+      builder.getFunctionType(TypeRange{anyOpType}, TypeRange{anyOpType});
+
+  auto newNamedSequence = builder.create<transform::NamedSequenceOp>(
+      loc, kKernelConfigSpecName, TypeAttr::get(seqType),
+      /*sym_visibility=*/StringAttr{},
+      /*arg_attrs=*/ArrayAttr{},
+      /*res_attrs*/ ArrayAttr{});
+
+  bool hasConsumedArg =
+      llvm::any_of(foreachMatchOps, [](transform::ForeachMatchOp op) {
+        Value operand = op->getOperand(0);
+        if (auto blockArg = mlir::dyn_cast<BlockArgument>(operand)) {
+          Operation *parentOp = blockArg.getOwner()->getParentOp();
+          if (auto namedSequenceOp =
+                  mlir::dyn_cast<transform::NamedSequenceOp>(parentOp)) {
+            return namedSequenceOp.getArgAttr(blockArg.getArgNumber(),
+                                              kArgConsumedAttrName) != nullptr;
+          }
+        }
+        return false;
+      });
+
+  StringRef attrName =
+      hasConsumedArg ? kArgConsumedAttrName : kArgReadOnlyAttrName;
+  newNamedSequence.setArgAttr(0, attrName, builder.getUnitAttr());
+  newNamedSequence->setAttr(kTuningSpecEntrypointAttrName,
+                            builder.getUnitAttr());
+  // Indicate the output module is a default tuning spec after merging.
+  module->setAttr(kTuningSpecDefaultEntrypointAttrName, builder.getUnitAttr());
+
+  // Step 3-C: Create a new block inside the NamedSequenceOp and merging
+  // ForeachMatchOp from each inner modules into one ForachMatchOp.
+  SmallVector<Type, 4> resultTypes;
+  llvm::append_range(resultTypes, expectedResultTypes);
+
+  SmallVector<std::pair<SymbolRefAttr, SymbolRefAttr>> matcherActionPairs;
+  SmallVector<Value, 4> forwardedInputs;
+  for (auto foreachMatchOp : foreachMatchOps) {
+    ArrayAttr matchers = foreachMatchOp.getMatchers();
+    ArrayAttr actions = foreachMatchOp.getActions();
+
+    for (size_t i = 0; i < matchers.size(); i++) {
```

Comment:
Do not re-evaluate the end iterator: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop